Ensure function parameter names always matches parameter types length

Marenz opened the pull request.

Marenz committed:
Ensure function parameter names always matches parameter types length

Marenz committed:
Remove duplicate if-check

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1234,10 +1234,10 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
 					if (!dynamic_cast<RationalNumberType const&>(*types[i]).mobileType())
 						m_errorReporter.fatalTypeError(components[i]->location(), "Invalid rational number.");
 
-				if (_tuple.isInlineArray())
```
Haha, nice one! :)

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1050,6 +1050,13 @@ class FunctionType: public Type
 		m_declaration(_declaration)
 	{
+		if (m_parameterNames.empty() && !m_parameterTypes.empty())
+			m_parameterNames = strings(m_parameterTypes.size(), "");
```
I would prefer to do this filling in the special-purpose constructor above only ("`Function type constructor to be used for a plain type (not derived from a declaration).`") and add the assertion here.

Marenz replied:
I am now working on that. So far it requires fixes all over the code because both c'tors are used

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1678,17 +1678,10 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
-			);
-			return;
-		}
+		solAssert(
```
Where is the check that allows you to turn this into an assertion?

Marenz replied:
If we can assume that the parameter names list is the same lengths as the parameter types list, which we can with the newly added asserts in the c'tor, then line 1593 does this check already: https://github.com/ethereum/solidity/pull/5957/files#diff-9f2a45efaa7b75ab7ec620d3177cf01aL1593

Marenz commented:
Updated:
* Added the same assert to all "final" c'tors
* Fixed code all over to make sure the lengths are correct
* Also added the fix for the *return* parameter name list

Marenz commented:
Update: Hopefully fixed the CI failure
